pilot/pkg/networking/core/v1alpha3/listener_builder.go

```diff
@@ -145,6 +147,13 @@ func (builder *ListenerBuilder) aggregateVirtualInboundListener() *ListenerBuild
 	// TODO: Trim the inboundListeners properly. Those that have been added to filter chains should
 	// be removed while those that haven't been added need to remain in the inboundListeners list.
 	filterChains, needTLS := reduceInboundListenerToFilterChains(builder.inboundListeners)
+	sort.SliceStable(filterChains, func(i, j int) bool {
+		if filterChains[i].Metadata != nil && filterChains[j].Metadata != nil {
+			return filterChains[i].Metadata.FilterMetadata[PilotMetaKey].String() <
+				filterChains[j].Metadata.FilterMetadata[PilotMetaKey].String()
```

Comment:
Rather than doing this, can you please use the name field in the filter chain? @silentdai added support for it during 1.3 cycle. So all envoys should have it. Otherwise, when we move to the named filter chains, we will have a chance of this regression once again.

Reply:
Also can we add a test to run this builder multiple times and validate every time it returns the same order - kind of simulating what Envoy does? That way, any changes in future probably we can catch with that test

Reply:
The name field is here: https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/listener/listener.proto#L204

Reply:
> The name field is here: https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/listener/listener.proto#L204

Thanks @lambdai . I'm struggling with coming up with a good naming scheme which incorporates all the variations due to filter chain match clauses.  The suggestion in here: https://github.com/istio/istio/pull/18091#issuecomment-544778945 doesn't really work as we are trying to sort filter chains which are already in a listener but differ in match clauses and possibly no match clause for default.